### PR TITLE
Ahmed/develop

### PR DESCRIPTION
**Pull request by AhmedMugtaba**
## Description of the change

Updates the main output graph legend to allow toggling specific curves. 

Screen Shot 2020-04-16 at 6 00 57 PM

Ongoing things to resolve:
- CurveData values are not calculated on page load.

Screen Shot 2020-04-16 at 4 26 15 PM

Screen Shot 2020-04-16 at 3 54 07 PM

## Type of change

- [x] New feature (interactive legend. when legend button cliked curveData will toggle certain metric like exposed, hospitalized, etc)

## Related issues

Closes #23 

### Development

These boxes should be checked by the submitter prior to merging:

- [x] Manual testing has been performed locally

### Code review

These boxes should be checked by reviewers prior to merging:

- [x] This pull request has a descriptive title and information useful to a reviewer
- [x] This pull request has been moved out of a Draft state, has no "Work In Progress" label, and has assigned reviewers
- [x] Potential security implications or infrastructural changes have been considered, if relevant

### CODE REVIEW

File: src/hooks/useToggle.tsx [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+  const [state, setState] = useState(true) as any;
+
+  const toggle = () => {
+    setState(!state);
+  };
+
+  return [state, toggle];
+}
+
+export default useToggle;
```
Comment (macfarlandian) [inline src/hooks/useToggle.tsx:13]: you don't seem to actually be using this anywhere? if not needed please delete

File: src/impact-dashboard/ChartArea.tsx [hunk 2]
```diff
@@ -25,22 +36,74 @@ const LegendContainer = styled.div`
...
+      filteredGroupStatus.reduce(
+        (data, key) => Object.assign(data, { [key]: curveData[key] }),
+        {},
+      ),
+    );
+  };
+
+  useEffect(() => {
+    updateDisplayData();
+  }, [groupStatus]);
```
Comment (macfarlandian) [inline src/impact-dashboard/ChartArea.tsx:93]: as we discussed earlier, it would be preferable for the data processing to remain inside `CurveChartContainer` for now. Why can't this logic run inside `CurveChartContainer` if you were to pass it `groupStatus` as a prop instead of the entire `curveData` structure?

File: src/impact-dashboard/CurveChartLegend.tsx [hunk 3]
```diff
@@ -15,19 +15,49 @@ const LegendItem = styled.div`
...
     <LegendWrapper>
-      <LegendItem color={markColors.exposed}>exposed</LegendItem>
-      <LegendItem color={markColors.infectious}>infectious</LegendItem>
-      <LegendItem color={markColors.hospitalized}>hospitalized</LegendItem>
-      <LegendItem color={markColors.fatalities}>fatalities</LegendItem>
+      <LegendItem
+        onClick={() => toggleGroup("exposed")}
+        color={markColors.exposed}
+      >
+        {groupStatus["exposed"] ? "exposed" : "exposed"}
```
Comment (macfarlandian) [inline src/impact-dashboard/CurveChartLegend.tsx:41]: this logic (and similar statements for the other items) isn't doing anything, you can just render the string directly

File: index.html [hunk 4]
```diff
@@ -1,10 +1,10 @@
-<!--
+<!-- 
   generatePageContent.tsx looks for the following strings in this file:
   - "TITLE PLACEHOLDER"
   - "<style></style>"
   - "CONTENT PLACEHOLDER"
-  If you change any of these, make sure to edit generatePageContent.tsx as well.
- -->
+  If you change any of these, make sure to edit generatePageContent.tsx as well. -->
+
```
Comment (macfarlandian) [inline index.html:7]: this looks like junk, I would just revert this

File: src/page-overview/OverviewPage.tsx [hunk 5]
```diff
@@ -1,3 +1,4 @@
+import { isIE } from "react-device-detect";
```
Comment (macfarlandian) [inline src/page-overview/OverviewPage.tsx:1]: I think it would be better if you just reverted the IE detection commit to get rid of these changes? that way we're not looking at two different features here

**Comment by jessex** [comment]
Hey @AhmedMugtaba that PR description is a bit hard to follow. Would you mind taking another pass at updating it so it's focused on a specific task and references a specific issue? Feel free to take a look at other PRs in this repository to get a sense for what's expected here.

File: src/page-overview/OverviewPage.tsx (hunk 5 above)
Comment (jessex) [reply src/page-overview/OverviewPage.tsx:1]: Agreed. For the record, @AhmedMugtaba, we want each PR to be focused on a single task, and we don't want to include work that we've already decided we're no longer pursuing. You should create a new branch each time you begin work on a new feature or task, as opposed to one large branch that you work on across all of your features.

File: src/impact-dashboard/CurveChartLegend.tsx (hunk 3 above)
Comment (jessex) [reply src/impact-dashboard/CurveChartLegend.tsx:41]: What's the idea here, @AhmedMugtaba? Why have a ternary operator that has the same value on both sides?

File: src/impact-dashboard/CurveChart.tsx [hunk 6]
```diff
@@ -137,6 +137,7 @@ const CurveChart: React.FC<CurveChartProps> = ({
   curveData,
   hospitalBeds,
   markColors,
+  // toggled,
```
Comment (jessex) [inline src/impact-dashboard/CurveChart.tsx:140]: We should remove this before merging, unless we actually need it.

File: src/impact-dashboard/ChartArea.tsx [hunk 7]
```diff
@@ -25,22 +36,74 @@ const LegendContainer = styled.div`
   flex: 0 0 auto;
 `;
 
-const ChartArea: React.FC = () => {
+function combinePopulations(data: CurveData, columnIndex: number) {
```
Comment (jessex) [inline src/impact-dashboard/ChartArea.tsx:39]: This point might be moot if you follow @macfarlandian's advice to move the calculation logic back to where it was previously. But if it does get moved out for some reason, we should try to not have two identical definitions of this function in two separate places.

File: src/impact-dashboard/ChartArea.tsx (hunk 2 above)
Comment (AhmedMugtaba) [reply src/impact-dashboard/ChartArea.tsx:93]: I think it's better for the parent component to hold the data, makes it a lot easier to share data between different children, instead of sibling communication, I put this data initially on it's own context. but as per our discussion, that was overkill. but you're are totally right it may seem like not needed at the moment but it will things a lot easier going forward.

File: src/impact-dashboard/CurveChartLegend.tsx (hunk 3 above)
Comment (AhmedMugtaba) [reply src/impact-dashboard/CurveChartLegend.tsx:41]: This was meant to be a place holder to implement Active/Inactive functionality on legend when the user toggled, will update today.

File: index.html (hunk 4 above)
Comment (AhmedMugtaba) [reply index.html:7]: Agree, will update.

File: src/page-overview/OverviewPage.tsx (hunk 5 above)
Comment (AhmedMugtaba) [reply src/page-overview/OverviewPage.tsx:1]: Got it, will update.

File: src/impact-dashboard/ChartArea.tsx (hunk 7 above)
Comment (AhmedMugtaba) [reply src/impact-dashboard/ChartArea.tsx:39]: this function is not needed here, just forgot to delete it, but I think it's better for the parent component to hold the data, makes it a lot easier to share data between different children, instead of sibling communication, I put this data initially on it's own context. but as per my discussion with Ian, it may seem like not needed at the moment but it will make things a lot easier going forward.

File: src/impact-dashboard/CurveChart.tsx (hunk 6 above)
Comment (AhmedMugtaba) [reply src/impact-dashboard/CurveChart.tsx:140]: Agree, will update.

File: src/hooks/useToggle.tsx (hunk 1 above)
Comment (AhmedMugtaba) [reply src/hooks/useToggle.tsx:13]: this left here as a handy toggle hook, in case we needed to use it in other parts of the App, but will delete and update again when needed.

File: src/impact-dashboard/ChartArea.tsx (hunk 2 above)
Comment (macfarlandian) [reply src/impact-dashboard/ChartArea.tsx:93]: ok but, the curve data isn't being shared between siblings. the only thing that's being shared is the toggle status. The chart should own the curve data because it is specific to what the chart displays. 

There is an argument to be made that the full projection data should be shared to avoid doing those calculations twice — in fact I left some TODOs to that effect so we could think about coming back to clean that up later — but that's not what you're doing here. IMO this is just unnecessarily creating a tight coupling between the chart component and its parent.

Comment (AhmedMugtaba) [reply src/impact-dashboard/ChartArea.tsx:93]: Alright, will update.

**Comment by macfarlandian** [comment]
hey @AhmedMugtaba this is very close! I think I have figured out why the curves are not working on initial page load; I'm going to just merge this into a temporary branch and take a stab at fixing it while you move on to your next ticket.
